Lib/xml/etree/ElementTree.py

```diff
@@ -1239,18 +1241,23 @@ def iterator(source):
                 pullparser.feed(data)
             root = pullparser._close_and_return_root()
             yield from pullparser.read_events()
-            it.root = root
+            it = wr()
+            if it is not None:
+                it.root = root
         finally:
             if close_source:
                 source.close()
 
     class IterParseIterator(collections.abc.Iterator):
         __next__ = iterator(source).__next__
-    it = IterParseIterator()
-    it.root = None
-    del iterator, IterParseIterator
 
-    next(it)
+        def __del__(self):
+            if close_source:
+                source.close()
+
+    it = IterParseIterator()
+    wr = weakref.ref(it)
+    del IterParseIterator
     return it
```

Comment:
I am curious, why previously both `iterator` and `IterParseIterator` names were deleted, but now only `IterParseIterator`? And what is the purpose of this statement in the first place? I was thinking that `iterator.__closure__` stores references to these objects, but as per my checks, closure stores only referenced variables inside, `pullparser`, 'close_source` and `wr` in this case.